### PR TITLE
Basic game engine

### PR DESCRIPTION
**Pull request by Michael-F-Bryan**
After having a look at what you said on the [forums] and the C++ reference implementation's [tutorial], I've stubbed out the basics for the front-end game engine API. I also included an example to see what it'll feel like from an end-user's point of view. 

What are your thoughts on what I've done so far?

[forums]: https://users.rust-lang.org/t/mentorship-on-a-side-project/12900/4?u=michael-f-bryan
[tutorial]: https://github.com/Blizzard/s2client-api/blob/master/docs/tutorial2.md

### CODE REVIEW

File: src/bot.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+pub trait Bot {
+    fn tick(&mut self, game_state: &GameState) -> ActionSet;
+    fn game_start(&mut self) {}
+    fn game_end(&mut self) {}
+}
+
+impl<F> Bot for F
+where
+    F: FnMut(&GameState) -> ActionSet,
```
Comment (ttdonovan) [inline src/bot.rs:11]: I'm a little bit confused by these few lines. Is this saying a the user of the lib can implement a there own definition of the the `tick()` function does for their own bot? Where `tick()` is a mutable function that accepts a reference to `GateState` and must return an `ActionSet`?

File: src/bot.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+impl ActionSet {
+    pub fn empty() -> ActionSet {
+        ActionSet::default()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.actions.is_empty()
+    }
+
+    pub(crate) fn actions(&self) -> ::std::slice::Iter<Action> {
```
Comment (ttdonovan) [inline src/bot.rs:34]: I've not seen `pub(crate)` before what does this do?

**Review by ttdonovan — COMMENTED** [review]
@Michael-F-Bryan Thank you for taking the time to layout a foundation for the API. Not sure if I would have been able to come to this conclusion as quickly as you did. It also appears you were able to correct my issues with `error_chain` and the `ws` returned results. I was not clear to me what I needed to do to get error_chain working. Is this a common pattern found in most crates where you'll find a `type Result<T> = StdResult<T, Error>;` that can be consumed by error_chain?

File: src/bot.rs (hunk 1 above)
Comment (Michael-F-Bryan) [reply src/bot.rs:11]: Yep, it means that you can use a closure as a bot, like I did in `examples/simple_bot.rs`. It's mainly for convenience or simple bots where you just want to do something every tick and don't care about the start and end lifecycle methods.

File: src/bot.rs (hunk 2 above)
Comment (Michael-F-Bryan) [reply src/bot.rs:34]: It means that this thing is `pub`, but only crate-wide. So then *we* can access the `actions()` easily, but it's still hidden from general users.

[relevant rfc](https://github.com/nox/rust-rfcs/blob/master/text/1422-pub-restricted.md)

**Comment by Michael-F-Bryan** [comment]
After looking through the C++ reference API more deeply I think I've misunderstood the scale of things. You'll usually execute an action every few frames instead of many actions every frame. As such, `ActionSet` doesn't really make sense and the `on_tick()` method should return an `Option<Action>`. 

I'll probably make a new PR when I get back from work which removes `ActionSet` altogether.

**Pull request merged**